### ISSUE TITLE
Fix QuickBooks payment method not working due to core changes

### ISSUE THREAD

**Issue by Pebblo** [eventespresso/event-espresso-core#3906]
This commit: https://github.com/eventespresso/event-espresso-core/commit/c87229c74c1c152889c8b0838dcd77b22a255737

And then the changes in #3806 

Have broken the Quickbooks payment method, it can no longer connect to Quickbooks.

Current (broken) payment method settings: 
![%E2%80%B9+EE4+Beaver+Bulder+%E2%80%94+WordPress+-+Google+Chrome+20](https://user-images.githubusercontent.com/1751030/176483931-39eb948f-cb8b-482c-8e91-dcb12e326cbd.png)

Clicking to connect shows this:
![Untitled+-+Google+Chrome+2022-06-23+14 30 38](https://user-images.githubusercontent.com/1751030/176484077-8f0fc1c6-2c10-4657-9b6d-279d1e22e7b3.png)

---

V4.10.29 does this:
![%E2%80%B9+EE4+Beaver+Bulder+%E2%80%94+WordPress+-+Google+Chrome+20 (1)](https://user-images.githubusercontent.com/1751030/176484177-1fb0c49b-38ec-49af-9abe-881a007b56d2.png)

Clicking to connect does this:
![Intuit+Accounts+-+Sign+In+-+Google+Chrome+2022-06-](https://user-images.githubusercontent.com/1751030/176484261-cffba15c-ee52-4936-a065-0b90765c4e5d.png)

**Comment by joaovictorangeline** [comment]
Related ticket: https://secure.helpscout.net/conversation/1927505799/908690/

**Comment by garthkoyle** [comment]
I let the customer know they could try again.

**Comment by garthkoyle** [comment]
On Event Smart, there is a console error: `bad request` when trying to disconnect:

```var qb_error = 'oAuth request Error.';
                        EEA_QUICKBOOKS_OAUTH.ee_ajax_loader.fadeOut('fast');
                        if (response['qb_error'])
                            qb_error = response['qb_error'];
                        console.log(qb_error);```

**Comment by hoseinrafiei** [comment]
@garthkoyle

I tested on master branch and I was able to disconnect without any problem.
I need to check on ES. Will contact you for credentials.

**Comment by garthkoyle** [comment]
You have super admin access now @hoseinrafiei

**Comment by garthkoyle** [comment]
@hoseinrafiei did you get a chance to look at why Quickbooks wouldn’t disconnect on Event Smart? https://lwhansonpc.eventsmart.com/wp-admin/admin.php?page=espresso_payment_settings&action=default&payment_method=quickbooks_onsite. It looks like the event organizer deactivated QB, but if you re-activate it and then try disconnecting you’ll see it doesn’t work.

**Comment by hoseinrafiei** [comment]
@garthkoyle

Yesterday I checked a little but didn't find anything.
Right now, I'm working on another high priority task and after closing this one, I will work again on the Quickbooks issue.

**Comment by hoseinrafiei** [comment]
@tn3rb 

Based on my checks the `Bad Request` with 400 as error code comes from the `revoke` command of QuickBooks API.
````
https://developer.api.intuit.com/v2/oauth2/tokens/revoke
````
in my local it happens whenever we revoked the access and we send new requests to revoke it. You can replicate it by inserting an exit immediately after the `wp_remote_post` function.
<img width="959" alt="Screen Shot 2022-06-29 at 18 20 00" src="https://user-images.githubusercontent.com/29144542/176453212-4d0bdf6f-f5b4-4148-86da-f32810ab270f.png">

The first time, you will receive a `200 OK` response but since the rest of code did but second time you will receive a `400 Bad Request` response.
Another way to replicate this issue is to change one of the Client ID or Client Secret fields after connection to the Quickbooks.

In my idea we can do the followings:

1- Make the Client ID and Client Secret fields read-only after connection. To change the credentials, client should disconnect first and then modify the fields.
2- We may run our revoke logic even if a `400 Bad Request` response received from Quickbooks which means something is wrong with the connection or the connection is already revoked.

What's your idea?

**Comment by tn3rb** [comment]
I don't doubt your findings regarding the `400 Bad Request` response, but that doesn't explain that @Pebblo found specific commits in the git log as being where the problems began. I'm assuming he performed a git bisect (or just walked back through the commit log manually) to find the point where the QuickBooks issues began.  So if commits A and B are fine, and then things break at commit C, we can only assume that something in commit C must be responsible, even if the code seems completely unrelated. 

This is what you will unfortunately need to do:

- checkout the last commit prior to the ones that @Pebblo pointed out
- verify that the Quickbooks (QB) payment method (PM) can connect and disconnect
- slowly, one by one, apply your changes from https://github.com/eventespresso/event-espresso-core/commit/c87229c74c1c152889c8b0838dcd77b22a255737 and  https://github.com/eventespresso/event-espresso-core/pull/3806 
- test the QB PM after every change until you find the one that causes the breakage
- figure out where/how the escaping is affecting the data used by QB and add the appropriate tags/attributes to your `AllowedTags` class (or wherever is most appropriate to do so)

with regards to:

>  my idea we can do the followings:
>  
>  1- Make the Client ID and Client Secret fields read-only after connection. To change the credentials, client should disconnect first and then modify the fields.
>  2- We may run our revoke logic even if a 400 Bad Request response received from Quickbooks which means something is wrong with the connection or the connection is already revoked.
 
both of those ideas sound logical to me, but you will have to run that past @knazart since he is our gateways expert and will know more about the situation than i do. I'll ping @knazart in chat regarding the above in order to get his feedback.

**Comment by tn3rb** [comment]
@hoseinrafiei 
with regards to the todo bullet list above, i forgot to add that (as always) I'm open to other ideas, approaches, whatever as long as they address the fact that something in from https://github.com/eventespresso/event-espresso-core/commit/c87229c74c1c152889c8b0838dcd77b22a255737 or https://github.com/eventespresso/event-espresso-core/pull/3806 has broken the QuickBooks PM, even if those changes seem completely unrelated.

**Comment by knazart** [comment]
> Make the Client ID and Client Secret fields read-only after connection. To change the credentials, client should disconnect first and then modify the fields

For now, yes, definitely. The issue here is that these params should not be provided by the merchant, but by our middle-man server. This is the part that needs refactoring.

> 2- We may run our revoke logic even if a 400 Bad Request response received from Quickbooks which means something is wrong with the connection or the connection is already revoked.

At first, checking for the `oAuth request Error` error if the response code is not `200` was enough. This was based on the types of errors this API was able to return. But from our experience this surely doesn't seem to work well enough. 
So maybe we should just reset the connection data without any conditionals, when Disconnect is triggered. And then send the API request checking the response to simply log if it's not `200`.

**Comment by hoseinrafiei** [comment]
@tn3rb

I checked out the https://github.com/eventespresso/event-espresso-core/commit/efa09ad0d94b96876a073a107ff11f1e3b371c9a commit which was prior to my adjustments and tested the connection and disconnection.

The issue happened on that version too. I wasn't able to disconnect using the exact method that I used on HEAD revision of master branch so I believe the disconnection problem was there prior my changes. We just detected it recently. Despite of that, Tony's report is related to the connection issue which we already fixed and he didn't mention disconnection at all.

Long story short, based on my checks the same issue can be replicated on the https://github.com/eventespresso/event-espresso-core/commit/efa09ad0d94b96876a073a107ff11f1e3b371c9a revision too.

I implemented my two suggestions and you can review it now: https://github.com/eventespresso/eea-quickbooks-gateway/pull/5
The first improvements is to fix the disconnection issue by running EE revoke logic even if we received `400` code from QuickBooks API and the second one is to disable the API credential fields in order to avoid modifications when we're already connected.

@knazart
Thanks for your feedback. Please check the new PR and let me know your feedback.
